### ISSUE TITLE
[A11Y] When an image links to content, the alt tag should be the title of the destination

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#5765]
## Description of the bug

When a screenreader is navigating a page and comes across an image that links somewhere, the screenreader needs to be able to explain where the link will take the navigator. 

This is commonly done by changing the `alt` tag on the image from a description of the image, to a description of the destination.

If the image is purely decorative, and is not necessary for navigation, another option would be to add a `tabindex=-1` on the link to skip it entirely for screenreaders. (If the post also has a text heading that links to the content, for example, it would be safe to skip the image for screenreaders)

Since the CMS cannot know when it is safe to skip the image (as in a teaser display mode) and when it is not safe to skip the image (as in a view, where the image/link is the only navigation) I think it would be safer to always add the alt text. This would ensure that all Backdrop sites have all the necessary navigation tools OOTB, and if people prefer to hide some elements to improve screen reader navigation, they can adjust their own sites further. 

I believe we should incorporate this change into [theme_image_formatter()](https://docs.backdropcms.org/api/backdrop/core%21modules%21image%21image.theme.inc/function/theme_image_formatter/1).

If we passed along an optional ALT tag as a 4th parameter, we could use that to override the image's specified ALT tag if provided.

For example, in [image_field_formatter_view()](https://docs.backdropcms.org/api/backdrop/core%21modules%21image%21image.field.inc/function/image_field_formatter_view/1)(
```php
    $element[$delta] = array(
      '#theme' => 'image_formatter',
      '#item' => $item,
      '#image_style' => $display['settings']['image_style'],
      '#path' => isset($uri) ? $uri : '',
      '#alt' => $entity->label(),
    );
```

and in [theme_image_formatter()](https://docs.backdropcms.org/api/backdrop/core%21modules%21image%21image.theme.inc/function/theme_image_formatter/1):
```php
function theme_image_formatter($variables) {
  $item = $variables['item'];

  // Do not output an empty 'title' attribute.
  if (isset($item['title']) && backdrop_strlen($item['title']) == 0) {
    unset($item['title']);
  }

  // Override the alt tag with destination information if the image is being used as a link. 
  if (isset($variables['path']['path']) && isset($variables['alt'])) {
    $item['alt'] = $variables['alt'];
  }

  if ($variables['image_style']) {
    $item['style_name'] = $variables['image_style'];
    $output = theme('image_style', $item);
  }
  else {
    $output = theme('image', $item);
  }

  // The link path and link options are both optional, but for the options to be
  // processed, the link path must at least be an empty string.
  if (isset($variables['path']['path'])) {
    $path = $variables['path']['path'];
    $options = isset($variables['path']['options']) ? $variables['path']['options'] : array();
    // When displaying an image inside a link, the html option must be TRUE.
    $options['html'] = TRUE;
    $output = l($output, $path, $options);
  }

  return $output;
}
```

Resources: 
* https://www.w3.org/WAI/tutorials/images/functional/

**Comment by indigoxela** [comment]
As expected, a failing test: "Image linked to file formatter displaying correctly on full node view."

So this needs a little more work.

TBH, I'm not sure if it's safe to make that change.

**Comment by jenlampton** [comment]
I've got the tests passing (they caught a bug, thank you tests!) 

@indigoxela what do you think is unsafe about adding an alt tag to linked images? Are there cases where that could be a problem?

**Comment by indigoxela** [comment]
> what do you think is unsafe about adding an alt tag to linked images? Are there cases where that could be a problem?

@jenlampton that's what I'm not yet sure of. I agree with the main intention here, but there could be situations, where this changed behavior leads to unexpected results, so this needs thorough testing and consideration of different scenarios.

**Comment by indigoxela** [comment]
Re failing tests: I can see that two different tests fail once in different php versions - it might be that at least one, if not both, are random failures.

@jenlampton I just watched the dev meeting and realized, that you're completely aware of all the possible culprits.

Two examples, for people who just chime in and are new to the accessibility topic:

**Example 1**
A view of node teasers, with title (linked), image (linked) and a "read more" link - so you end up the three links pointing to exactly the same target, but all three have different text. The title provides only the node title, the image provides the image description (which doesn't say anything about where the link leads to), the "read more" link has the most accessible text (for assistive technologies, partly visually hidden) "Read more about Whatever title the node has".

Now, for better A11Y you (site builder) should remove the two _redundant_ links, but then the sighted users will be baffled (neither the title, nor the image are links, very uncommon). In this case `tabindex=-1` will work fine for the redundant link around the image. But what about the still redundant title (teaser heading is also a link)?

**Example 2**
A view of only images pointing to nodes - the alt attribute text is the only hint for screen reader users, what "this link thing means", but usually the alt text describes the image. In this case it should describe where the link leads to instead.

This is all pretty complex. The field formatter itself doesn't "know" where it is, what the context of the image is at that point, and so it's tricky to guess when to touch the alt text and when not.

**Comment by BWPanda** [comment]
I'm hesitant about this part:

> we could use that to **override** the image's specified ALT tag

Instead, would it not be better to *append* something like "Links to [node:title]" to the existing alt text...?

**Comment by klonos** [comment]
I share the same concern with @BWPanda on this one @jenlampton. If the alt attribute contains automated text that was not explicitly/intentionally specified by content editors, then that may be OK, but if we are "automagically" replacing text that was intentionally added as alt, then that may be WTF.

I like @BWPanda's suggestion to be _appending_ information about the destination, but not completely overriding/replacing the existing alt.
